charms/tensorboard-controller/tests/integration/test_charm.py

```diff
@@ -0,0 +1,200 @@
+    istio_pilot = "istio-pilot"
+    await _setup_istio(ops_test, istio_gateway, istio_pilot)
+
+    # add Tensorboard-Controller/Istio relation
+    await ops_test.model.add_relation(f"{istio_pilot}:gateway-info", f"{APP_NAME}:gateway-info")
+
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME], status="active", raise_on_blocked=True, timeout=1000
+    )
+
+
+@tenacity.retry(
+    wait=tenacity.wait_exponential(multiplier=2, min=1, max=10),
+    stop=tenacity.stop_after_attempt(30),
+    reraise=True,
+)
+def assert_replicas(client, resource_class, resource_name, namespace):
+    """Test for replicas.
+
+    Retries multiple times to allow for tensorboard to be created (i.e. readyReplicas == 1).
+    """
+    tensorboard = client.get(resource_class, resource_name, namespace=namespace)
+    replicas = tensorboard.get("status", {}).get("readyReplicas")
+
+    resource_class_kind = resource_class.__name__
+    if replicas == 1:
+        logger.info(f"{resource_class_kind}/{resource_name} readyReplicas == {replicas}")
+    else:
+        logger.info(
+            f"{resource_class_kind}/{resource_name} readyReplicas == {replicas} (waiting for '1')"
+        )
+
+    assert replicas == 1, f"Waited too long for {resource_class_kind}/{resource_name}!"
+
+
+async def test_create_tensorboard(ops_test: OpsTest):
+    """Test Tensorboard creation."""
+    lightkube_client = Client()
+    load_in_cluster_generic_resources(lightkube_client)
+
+    # Create PVC for Tensorboard logs and Tensorboard
+    resources = codecs.load_all_yaml(
+        "\n---\n".join([PVC_TEMPLATE_FILE.read_text(), TENSORBOARD_TEMPLATE_FILE.read_text()]),
+        context={"pvc_name": PVC_NAME, "tensorboard_name": TENSORBOARD_NAME},
+    )
+    for rsc in resources:
+        lightkube_client.create(rsc, namespace=ops_test.model_name)
+
+    try:
+        tensorboard_ready = lightkube_client.get(
+            TENSORBOARD_RESOURCE,
+            name=TENSORBOARD_NAME,
+            namespace=ops_test.model_name,
+        )
+    except ApiError:
+        assert False
+    assert tensorboard_ready, f"Tensorboard {ops_test.model_name}/{TENSORBOARD_NAME} not found!"
+
+    assert_replicas(lightkube_client, TENSORBOARD_RESOURCE, TENSORBOARD_NAME, ops_test.model_name)
+
+
+@pytest.mark.abort_on_fail
+async def test_remove_with_resources_present(ops_test: OpsTest):
+    """Test remove with all resources deployed.
+
+    Verify that all deployed resources that need to be removed are removed.
+    """
+    # remove deployed charm and verify that it is deleted
+    await ops_test.model.remove_application(app_name=APP_NAME, block_until_done=True)
+    assert APP_NAME not in ops_test.model.applications
+
+    # verify that all resources that were deployed are deleted
+    lightkube_client = Client()
+
+    # verify that all created CRDs and Services in namespace are deleted
+    for rsc in [
+        CustomResourceDefinition,
+        Service,
```

Comment:
I'm unsure about checking for removal of the `Service`.  The reason I wonder about doing this is because it is entirely a juju operation (juju owns and manages this objects), so should we be checking for deletion?  If yes, where do we draw the line (there's lot of other things it does for us that we don't check)